homeassistant/components/nanoleaf/config_flow.py

```diff
@@ -0,0 +1,227 @@
+
+from .const import DOMAIN
+from .util import pynanoleaf_get_info
+
+_LOGGER = logging.getLogger(__name__)
+
+# For discovery integration import
+CONFIG_FILE: Final = ".nanoleaf.conf"
+
+USER_SCHEMA: Final = vol.Schema(
+    {
+        vol.Required(CONF_HOST): str,
+    }
+)
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Nanoleaf config flow."""
+
+    VERSION = 1
+
+    nanoleaf: Nanoleaf
+
+    # For discovery integration import
+    discovery_conf: dict
+    device_id: str
+
+    def __init__(self) -> None:
+        """Initialize a Nanoleaf flow."""
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle Nanoleaf flow initiated by the user."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=USER_SCHEMA, last_step=False
+            )
+        self._async_abort_entries_match({CONF_HOST: user_input[CONF_HOST]})
+        self.nanoleaf = Nanoleaf(user_input[CONF_HOST])
+        try:
+            await self.hass.async_add_executor_job(self.nanoleaf.authorize)
+        except Unavailable:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=USER_SCHEMA,
+                errors={"base": "cannot_connect"},
+                last_step=False,
+            )
+        except NotAuthorizingNewTokens:
+            pass
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unknown error connecting to Nanoleaf")
+            return self.async_show_form(
+                step_id="user",
+                data_schema=USER_SCHEMA,
+                last_step=False,
+                errors={"base": "unknown"},
+            )
+        return self.async_show_form(step_id="link")
+
+    async def async_step_zeroconf(
+        self, discovery_info: DiscoveryInfoType
+    ) -> FlowResult:
+        """Handle Nanoleaf Zeroconf discovery."""
+        _LOGGER.debug("Zeroconf discovered: %s", discovery_info)
+        return await self._async_discovery_handler(discovery_info)
+
+    async def async_step_homekit(self, discovery_info: DiscoveryInfoType) -> FlowResult:
+        """Handle Nanoleaf Homekit discovery."""
+        _LOGGER.debug("Homekit discovered: %s", discovery_info)
+        return await self._async_discovery_handler(discovery_info)
+
+    async def _async_discovery_handler(
+        self, discovery_info: DiscoveryInfoType
+    ) -> FlowResult:
+        """Handle Nanoleaf discovery."""
+        host = discovery_info["host"]
+        name = discovery_info["name"].replace(f".{discovery_info['type']}", "")
+        await self.async_set_unique_id(name)
+        self._abort_if_unique_id_configured({CONF_HOST: host})
+        self.nanoleaf = Nanoleaf(host)
+
+        # Import from discovery integration
+        self.device_id = discovery_info["properties"]["id"]
```

Comment:
Unfortunately zeroconf is not documented in the API docs.
This is set on my device (Model: Canvas) for both `_nanoleafms._tcp.local.` and `_nanoleafapi._tcp.local.`. 
We depend on this being present since https://github.com/home-assistant/core/pull/46407 (2021.4) in `_nanoleafapi._tcp.local.` ([netdisco](https://github.com/home-assistant-libs/netdisco/blob/master/netdisco/discoverables/nanoleaf_aurora.py)) and there have not been any know issues.

Reply:
With homekit, this value will change when the device is factory reset assuming they follow the homekit spec.  Since we aren't using it for the unique id, I think that is ok. 

Is that going to be a problem fetching the token?

Reply:
If you reset the device, the token is no longer valid and it should not be imported. The config flow will not import the device and continue as if it were a new device.